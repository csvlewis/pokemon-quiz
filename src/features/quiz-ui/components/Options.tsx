import { Button, Stack, Typography } from "@mui/material";
import { useQuizContext } from "@/features/context/components/QuizContextProvider";
import { setOptionColor } from "@/features/quiz-ui/utils/setOptionColor";
import { handleOptionClick } from "@/features/quiz-ui/utils/handleOptionClick";

export function Options(): React.ReactNode {
  const {
    currentQuestion,
    displayAnswer,
    questions,
    quizInProgress,
    selectedAnswer,
    setSelectedAnswer,
  } = useQuizContext();

  if (quizInProgress)
    return (
      <Stack spacing={2} direction={"row"}>
        {questions[currentQuestion - 1]?.options.map((option, index) => (
          <Button
            color={setOptionColor(option, displayAnswer)}
            variant={
              option.text === selectedAnswer.text ? "contained" : "outlined"
            }
            onClick={() =>
              handleOptionClick(
                option,
                displayAnswer,
                selectedAnswer,
                setSelectedAnswer
              )
            }
            key={index}
          >
            <Typography>{option.text}</Typography>
          </Button>
        ))}
      </Stack>
    );
}
