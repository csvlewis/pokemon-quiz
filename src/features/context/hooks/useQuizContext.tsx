import { useContext } from "react";
import { QuizContextType } from "@/features/context/types/QuizContextType";
import { QuizContext } from "@/features/context/components/QuizContext";

export function useQuizContext(): QuizContextType {
  return useContext<QuizContextType>(QuizContext);
}
