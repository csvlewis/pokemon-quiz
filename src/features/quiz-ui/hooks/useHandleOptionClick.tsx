import { useQuizContext } from "@/features/context/hooks/useQuizContext";
import { Option } from "@/types/Option";

export function useHandleOptionClick(): (option: Option) => void {
  const { displayAnswer, selectedAnswer, setSelectedAnswer } = useQuizContext();

  return (option: Option) => {
    if (displayAnswer) return;
    if (option === selectedAnswer) {
      setSelectedAnswer({});
    } else {
      setSelectedAnswer(option);
    }
  };
}
