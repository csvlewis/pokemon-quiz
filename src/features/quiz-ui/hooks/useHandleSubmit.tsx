import { useQuizContext } from "@/features/context/hooks/useQuizContext";
import { useIncrementQuestion } from "@/features/quiz-ui/hooks/useIncrementQuestion";

export function useHandleSubmit(): () => void {
  const {
    correctAnswers,
    displayAnswer,
    selectedAnswer,
    setCorrectAnswers,
    setDisplayAnswer,
  } = useQuizContext();
  const incrementQuestion = useIncrementQuestion();

  return () => {
    if (displayAnswer) {
      incrementQuestion();
    } else {
      if (selectedAnswer.isCorrect) setCorrectAnswers(correctAnswers + 1);
      setDisplayAnswer(true);
    }
  };
}
