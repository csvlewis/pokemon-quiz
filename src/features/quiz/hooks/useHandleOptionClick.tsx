import { Option } from "@/types/Option";
import { useQuizContext } from "@/features/context/hooks/useQuizContext";

export function useHandleOptionClick(): (option: Option) => void {
  const { displayAnswer, selectedAnswer, setSelectedAnswer } = useQuizContext();

  return (option: Option) => {
    if (displayAnswer) return;
    if (option === selectedAnswer) {
      setSelectedAnswer({ text: "", isCorrect: false });
    } else {
      setSelectedAnswer(option);
    }
  };
}
