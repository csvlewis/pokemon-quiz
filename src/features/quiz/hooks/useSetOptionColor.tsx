import { OverridableStringUnion } from "@mui/types";

import { Option } from "@/types/Option";
import { useQuizContext } from "@/features/context/hooks/useQuizContext";

export function useSetOptionColor(): (
  option: Option
) => OverridableStringUnion<"error" | "primary" | "success"> {
  const { displayAnswer } = useQuizContext();

  return (option: Option) => {
    let color: OverridableStringUnion<"error" | "primary" | "success"> =
      "primary";

    if (displayAnswer) {
      if (!option.isCorrect) {
        color = "error";
      } else {
        color = "success";
      }
    }

    return color;
  };
}
