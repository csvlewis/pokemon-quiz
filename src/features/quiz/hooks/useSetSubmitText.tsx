import { useQuizContext } from "@/features/context/hooks/useQuizContext";

export function useSetSubmitText(): () => string {
  const { currentQuestion, displayAnswer, questions } = useQuizContext();

  return () => {
    let text: string = "Submit";

    if (displayAnswer) {
      if (currentQuestion >= questions.length) {
        text = "See Results";
      } else text = "Next Question";
    }

    return text;
  };
}
