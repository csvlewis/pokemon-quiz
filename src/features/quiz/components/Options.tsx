import { OverridableStringUnion } from "@mui/types";
import { Button, Stack, Typography } from "@mui/material";
import { useQuizContext } from "@/features/context/hooks/useQuizContext";
import { useHandleOptionClick } from "@/features/quiz/hooks/useHandleOptionClick";
import { useSetOptionColor } from "@/features/quiz/hooks/useSetOptionColor";
import { Option } from "@/types/Option";

export function Options(): React.ReactNode {
  const { currentQuestion, questions, quizInProgress, selectedAnswer } =
    useQuizContext();

  const setOptionColor: (
    option: Option
  ) => OverridableStringUnion<"error" | "primary" | "success"> =
    useSetOptionColor();
  const handleOptionClick: (option: Option) => void = useHandleOptionClick();

  if (quizInProgress)
    return (
      <Stack spacing={2} direction={"row"}>
        {questions[currentQuestion - 1]?.options.map((option, index) => (
          <Button
            color={setOptionColor(option)}
            variant={
              option.text === selectedAnswer.text ? "contained" : "outlined"
            }
            onClick={() => handleOptionClick(option)}
            key={index}
          >
            <Typography>{option.text}</Typography>
          </Button>
        ))}
      </Stack>
    );
}
