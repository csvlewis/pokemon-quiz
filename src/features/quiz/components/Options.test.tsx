import React from "react";
import { Options } from "./Options";
import { render } from "@testing-library/react";
import { useQuizContext } from "@/features/context/hooks/useQuizContext";

jest.mock("@/features/context/hooks/useQuizContext");

const mockedUseQuizContext = useQuizContext as jest.Mock;

describe("Options", () => {
  it("should render correctly", () => {
    mockedUseQuizContext.mockReturnValue({
      currentQuestion: 1,
      questions: [
        {
          text: "Question 1",
          imageUrl: "https://example.com/image1.png",
          options: [
            { text: "Option 1", isCorrect: false },
            { text: "Option 2", isCorrect: true },
            { text: "Option 3", isCorrect: false },
            { text: "Option 4", isCorrect: false },
          ],
        },
      ],
      quizInProgress: true,
      selectedAnswer: { text: "Option 1", isCorrect: false },
    });

    const { container } = render(<Options />);
    expect(container).toMatchSnapshot();
  });
});
