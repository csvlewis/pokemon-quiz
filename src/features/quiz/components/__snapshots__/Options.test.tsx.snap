// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Options should render correctly 1`] = `
<div>
  <div
    class="MuiStack-root css-pl8gqh-MuiStack-root"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-74d805-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
      >
        Option 1
      </p>
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1blvszr-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
      >
        Option 2
      </p>
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1blvszr-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
      >
        Option 3
      </p>
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1blvszr-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
      >
        Option 4
      </p>
    </button>
  </div>
</div>
`;
